File: Java/src/_1046_Last_Stone_Weight/Solution.java
package _1046_Last_Stone_Weight;

import java.util.PriorityQueue;
import java.util.Queue;

/**
 * Date: 2020/12/30 9:31
 * Content:
 * 执行用时：2 ms, 在所有 Java 提交中击败了41.83%的用户
 * 内存消耗：36 MB, 在所有 Java 提交中击败了29.15%的用户
 */
public class Solution {

    public int lastStoneWeight(int[] stones) {

        Queue<Integer> heap = new PriorityQueue<>((o1, o2) -> o2 - o1);
        for (int x : stones) heap.add(x);
        while (heap.size() > 1) {
            int y = heap.remove(), x = heap.remove();
            if (x == y) continue;
            heap.add(y - x);
        }

        return heap.size() == 1 ? heap.remove() : 0;
    }

    public static void main(String[] args) {

        int[] stones = {2, 7, 4, 1, 8, 1};
        System.out.println((new Solution()).lastStoneWeight(stones));
    }
}

File: Python/_1046_Last_Stone_Weight/Solution.py
# coding=utf-8
# Date: 2021/6/8 10:11
from typing import List


# 执行用时：44 ms, 在所有 Python3 提交中击败了46.56%的用户
# 内存消耗：14.8 MB, 在所有 Python3 提交中击败了73.62%的用户
# python中只有小顶堆，需要将数据转为负数实现大顶堆操作
class Solution:
    def lastStoneWeight(self, stones: List[int]) -> int:
        from queue import PriorityQueue
        heap = PriorityQueue()
        for x in stones:
            heap.put(-x)
        while heap.qsize() > 1:
            y = heap.get(); x = heap.get()
            if x == y:
                continue
            heap.put(y - x)
        return -heap.queue[0] if heap.qsize() == 1 else 0


if __name__ == "__main__":
    print(Solution().lastStoneWeight([2, 7, 4, 1, 8, 1]))
